Congma/ck tile/tile engine streamk

CongMa13 opened the pull request:
## Proposed changes

Tile engine is ck tile version of ck profiler. 

    Add basic support of tile engine to streamk gemm in this PR. 
    - Only add streamk_gemm_instance_builder.py since the streamk is still
      under developing. Will add test_benchmark.sh and gemm_benchmark.py
      when streamk is stable.
    - It is able to generate the benchmark binaries separately.

## Checklist

Please put an `x` into the boxes that apply. You can also fill these out after creating the PR. If you're not sure, please don't hesitate to ask.

- [ ] I have added tests relevant to the introduced functionality, and the unit tests are passing locally
- [ ] I have added the test to REGRESSION_TESTS list defined at the top of CMakeLists.txt in tests/CMakeLists.txt, **IF** the test takes more than 30 seconds to run.
- [ ] I have added inline documentation which enables the maintainers with understanding the motivation
- [ ] I have removed the stale documentation which is no longer relevant after this pull request
- [ ] (If this change is user-facing) I have added release notes which provide the end users with a brief summary of the improvement from this pull request
- [x] I have run `clang-format` on all changed files
- [ ] Any dependent changes have been merged

## Discussion

If this is a relatively large or complex change, feel free to start a discussion by explaining why you chose the solution you did and what alternatives you considered

CongMa13 committed:
[CK TILE STREAMK] Tile engine supports streamk gemm

Add basic support of tile engine to streamk gemm.
- Only add streamk_gemm_instance_builder.py since the streamk is still
  under developing. Will add test_benchmark.sh and gemm_benchmark.py
  when streamk is stable.
- It is able to generate the benchmark binaries separately.

CongMa13 commented:
The PR is based on another PR 2781 which is not merged yet. So keep this PR a draft for now and will make it ready for review when 2781 is merged.

CongMa13 committed:
[CK TILE STREAMK] Tile engine supports streamk gemm

- Add tile engine streamk gemm into CI

cgmillette commented on tile_engine/ops/gemm_streamk/benchmark_gemm.hpp:
```diff
@@ -0,0 +1,242 @@
+}
+
+template <typename ADataType, typename BDataType, typename AccDataType, typename CDataType>
+auto calculate_rtol_atol(const ck_tile::index_t K,
```
This calculation seems to be repeated in many places within CK. Is there a common implementation?

CongMa13 replied:
The refactoring will be done in LWPCK-3885

cgmillette commented on tile_engine/ops/gemm_streamk/benchmark_gemm.hpp:
```diff
@@ -0,0 +1,242 @@
+
+inline std::string get_rocm_version()
+{
+    std::ifstream version_file("/opt/rocm/.info/version");
```
Is this information necessary? May want to search for an alternative API to this because this would not work on Windows.

CongMa13 replied:
Searched in documentation but did not find an alternative API.  Do you know any that works on Windows?

cgmillette commented on tile_engine/ops/gemm_streamk/benchmark_gemm.hpp:
```diff
@@ -0,0 +1,242 @@
+bool compare(std::string instanceName,
+             ck_tile::index_t K,
+             ck_tile::index_t kbatch,
+             ck_tile::HostTensor<CDataType>& c_m_n_dev_result,
```
Can we adjust the const-ness of the dev and host results? These should be not mutable

CongMa13 replied:
For sure. Changed.

cgmillette commented on tile_engine/ops/gemm_streamk/benchmark_gemm.hpp:
```diff
@@ -0,0 +1,242 @@
+
+/// @brief Function to get the kernel output with reference implementation on CPU/GPU
+void gemm_host_reference(int verify,
+                         ck_tile::HostTensor<ADataType>& a_m_k,
```
Here we may also need to adjust const-ness of the arguments.

CongMa13 replied:
For sure. Changed.

cgmillette commented on tile_engine/ops/gemm_streamk/benchmark_streamk_gemm_single.cpp:
```diff
@@ -0,0 +1,173 @@
+                "true",
+                "To flush cache, possible values are true or false. "
+                "Default is false.")
+        .insert("rotating_count", "1000", "number of iterations to rotate the cache. default is 5.")
```
Default doesn't appear to be 5 here

CongMa13 replied:
Changed to 1000.

cgmillette commented on tile_engine/ops/gemm_streamk/gemm_kernel_list.txt:
```diff
@@ -0,0 +1,16 @@
+gemm_fp8_rcr_mem_default_intrawave_False_False_False_False_256x256x32_4x1x1_32x32x32|256x256x32_4x1x1_32x32x32|mem_default_intrawave_false_false_false_false
```
Need this file?

CongMa13 replied:
This is a generated non-source file. Removed.

cgmillette commented on tile_engine/ops/gemm_streamk/gemm_kernel_count.txt:
```diff
@@ -0,0 +1 @@
+16
```
Need this file?

CongMa13 replied:
This is a generated non-source file. Removed.

CongMa13 commented on tile_engine/ops/gemm_streamk/gemm_kernel_count.txt:
should remove this

CongMa13 committed:
[CK TILE STREAMK] Enhance tile engine with streamk GEMM support

- Mark input parameters as `const` where applicable
- Remove generated non-source files

ecamartins commented on tile_engine/ops/gemm_streamk/CMakeLists.txt:
```diff
@@ -0,0 +1,295 @@
+    
+    # Create the executable
+    add_executable(${target_name} 
+        ${GEMM_SOURCE_DIR}/benchmark_streamk_gemm_single.cpp
```
Based on PR #3085, we likely want to add `EXCLUDE_FROM_ALL` here.

CongMa13 commented:
The streamk kernel has been reimplemented. Close this PR and reimplement the tile engine.
